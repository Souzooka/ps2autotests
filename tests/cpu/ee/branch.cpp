#include <stdio.h>
#include <tamtypes.h>

enum BranchResultFlags {
	BRANCH_SKIPPED = 0,
	BRANCH_FOLLOWED = 1,
	BRANCH_SET_RA = 2,
	BRANCH_DELAY_SLOT = 4,
};

static u32 __attribute__((aligned(16))) C_ZERO[4] = {0, 0, 0, 0};
static u32 __attribute__((aligned(16))) C_S16_MAX[4] = {0x7FFF, 0, 0, 0};
static u32 __attribute__((aligned(16))) C_S16_MIN[4] = {0xFFFF8000, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF};
static u32 __attribute__((aligned(16))) C_S32_MAX[4] = {0x7FFFFFFF, 0, 0, 0};
static u32 __attribute__((aligned(16))) C_S32_MIN[4] = {0x80000000, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF};
static u32 __attribute__((aligned(16))) C_S64_MAX[4] = {0xFFFFFFFF, 0x7FFFFFFF, 0, 0 };
static u32 __attribute__((aligned(16))) C_S64_MIN[4] = {0, 0x80000000, 0xFFFFFFFF, 0xFFFFFFFF};
static u32 __attribute__((aligned(16))) C_NEGONE[4] = {0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF};
static u32 __attribute__((aligned(16))) C_ONE[4] = {1, 0, 0, 0};
static u32 __attribute__((aligned(16))) C_GARBAGE1[4] = {0x1337, 0x1338, 0x1339, 0x133A};
static u32 __attribute__((aligned(16))) C_GARBAGE2[4] = {0xDEADBEEF, 0xDEADBEEE, 0xDEADBEED, 0xDEADBEEC};

#define BRRO_OP_FUNC(OP) \
static inline void BRRO_##OP(u32 &result, const register u128 &rs, const register u128 &rt) { \
	asm volatile ( \
		".set noreorder\n" \
		"" \
		"lui %0, 0\n" \
		"or $t9, $0, $ra\n" \
		"" \
		#OP " %1, %2, " #OP "_branch%=\n" \
		"ori %0, %0, 4\n" \
		"" \
		"b " #OP "_skip%=\n" \
		"nop\n" \
		"" \
		#OP "_branch%=:\n" \
		"ori %0, %0, 1\n" \
		"" \
		#OP "_skip%=:\n" \
		"beq $t9, $ra, " #OP "_done%=\n" \
		"nop\n" \
		"" \
		"or $ra, $0, $t9\n" \
		"ori %0, %0, 2\n" \
		"" \
		#OP "_done%=:\n" \
		"" \
		: "=&r"(result) : "r"(rs), "r"(rt) : "t9" \
	); \
}

#define BRO_OP_FUNC(OP) \
static inline void BRO_##OP(u32 &result, const register u128 &rs) { \
	asm volatile ( \
		".set noreorder\n" \
		"" \
		"lui %0, 0\n" \
		"or $t9, $0, $ra\n" \
		"" \
		#OP " %1, " #OP "_branch%=\n" \
		"ori %0, %0, 4\n" \
		"" \
		"b " #OP "_skip%=\n" \
		"nop\n" \
		"" \
		#OP "_branch%=:\n" \
		"ori %0, %0, 1\n" \
		"" \
		#OP "_skip%=:\n" \
		"beq $t9, $ra, " #OP "_done%=\n" \
		"nop\n" \
		"" \
		"or $ra, $0, $t9\n" \
		"ori %0, %0, 2\n" \
		"" \
		#OP "_done%=:\n" \
		"" \
		: "=&r"(result) : "r"(rs) : "t9" \
	); \
}

#define BO_OP_FUNC(OP) \
static inline void BO_##OP(u32 &result) { \
	asm volatile ( \
		".set noreorder\n" \
		"" \
		"lui %0, 0\n" \
		"or $t9, $0, $ra\n" \
		"" \
		#OP " " #OP "_branch\n" \
		"ori %0, %0, 4\n" \
		"" \
		"b " #OP "_skip\n" \
		"nop\n" \
		"" \
		#OP "_branch:\n" \
		"ori %0, %0, 1\n" \
		"" \
		#OP "_skip:\n" \
		"beq $t9, $ra, " #OP "_done\n" \
		"nop\n" \
		"" \
		"or $ra, $0, $t9\n" \
		"ori %0, %0, 2\n" \
		"" \
		#OP "_done:\n" \
		"" \
		: "=&r"(result) : : "t9" \
	); \
}

#define BRR_OP_FUNC(OP) \
static inline void BRR_##OP(u32 &result) { \
	asm volatile ( \
		".set noreorder\n" \
		"" \
		"lui %0, 0\n" \
		"la $t8, " #OP "_branch\n" \
		"or $t9, $0, $0\n" \
		"" \
		#OP " $t9, $t8\n" \
		"ori %0, %0, 4\n" \
		"" \
		"b " #OP "_skip\n" \
		"nop\n" \
		"" \
		#OP "_branch:\n" \
		"ori %0, %0, 1\n" \
		"" \
		#OP "_skip:\n" \
		"beq $t9, $0, " #OP "_done\n" \
		"nop\n" \
		"" \
		"ori %0, %0, 2\n" \
		"" \
		#OP "_done:\n" \
		"" \
		: "=&r"(result) : : "t8", "t9" \
	); \
}

#define BR_OP_FUNC(OP) \
static inline void BR_##OP(u32 &result) { \
	asm volatile ( \
		".set noreorder\n" \
		"" \
		"lui %0, 0\n" \
		"la $t8, " #OP "_branch\n" \
		"or $t9, $0, $ra\n" \
		"" \
		#OP " $t8\n" \
		"ori %0, %0, 4\n" \
		"" \
		"b " #OP "_skip\n" \
		"nop\n" \
		"" \
		#OP "_branch:\n" \
		"ori %0, %0, 1\n" \
		"" \
		#OP "_skip:\n" \
		"beq $t9, $ra, " #OP "_done\n" \
		"nop\n" \
		"" \
		"or $ra, $0, $t9\n" \
		"ori %0, %0, 2\n" \
		"" \
		#OP "_done:\n" \
		"" \
		: "=&r"(result) : : "t8", "t9" \
	); \
}

template <u32 i>
static inline void SET_U32(register u128 &rd) {
	asm volatile (
		"lui %0, %1\n"
		"ori %0, %0, %2\n"
		: "+r"(rd) : "K"((i >> 16) & 0xFFFF), "K"(i & 0xFFFF)
	);
}

static inline void SET_M(register u128 &rd, u32 *p) {
	rd = *(vu128 *)p;
}

static inline void PRINT_BRANCH(const u32 result, bool newline) {
	printf("%s, ", (result & BRANCH_FOLLOWED) != 0 ? "followed" : "skipped");
	printf("%s, ", (result & BRANCH_SET_RA) != 0 ? "set ra" : "ignored ra");
	printf("%s", (result & BRANCH_DELAY_SLOT) != 0 ? "ran delay slot" : "skipped delay slot");
	if (newline) {
		printf("\n");
	}
}

#define BRRO_OP_DO_II(OP, s, t) \
	SET_U32<s>(rs); SET_U32<t>(rt); \
	BRRO_##OP(res, rs, rt); \
	printf("  %s %d, %d: ", #OP, (u32)rs, (u32)rt); PRINT_BRANCH(res, true);

#define BRRO_OP_DO_MM(OP, s, t) \
	SET_M(rs, s); SET_M(rt, t); \
	BRRO_##OP(res, rs, rt); \
	printf("  %s %s, %s: ", #OP, #s, #t); PRINT_BRANCH(res, true);

#define TEST_BRRO_FUNC(OP) \
BRRO_OP_FUNC(OP); \
static void test_##OP() { \
	u32 res; \
	register u128 rs, rt; \
	 \
	printf("%s:\n", #OP); \
	 \
	BRRO_OP_DO_II(OP, 0, 0); \
	BRRO_OP_DO_II(OP, 0, 1); \
	BRRO_OP_DO_II(OP, 1, 1); \
	BRRO_OP_DO_II(OP, 1, 0); \
	BRRO_OP_DO_II(OP, 2, 2); \
	BRRO_OP_DO_II(OP, 0xFFFFFFFF, 1); \
	BRRO_OP_DO_II(OP, 0xFFFFFFFF, 0xFFFFFFFF); \
	 \
	BRRO_OP_DO_MM(OP, C_ZERO, C_ZERO); \
	BRRO_OP_DO_MM(OP, C_ZERO, C_ONE); \
	BRRO_OP_DO_MM(OP, C_ONE, C_ZERO); \
	BRRO_OP_DO_MM(OP, C_ONE, C_ONE); \
	BRRO_OP_DO_MM(OP, C_ONE, C_NEGONE); \
	BRRO_OP_DO_MM(OP, C_S16_MAX, C_S16_MAX); \
	BRRO_OP_DO_MM(OP, C_S16_MIN, C_S16_MIN); \
	BRRO_OP_DO_MM(OP, C_S32_MAX, C_S32_MAX); \
	BRRO_OP_DO_MM(OP, C_S32_MIN, C_S32_MIN); \
	BRRO_OP_DO_MM(OP, C_S64_MAX, C_S64_MAX); \
	BRRO_OP_DO_MM(OP, C_S64_MIN, C_S64_MIN); \
	BRRO_OP_DO_MM(OP, C_GARBAGE1, C_GARBAGE2); \
	\
	printf("\n"); \
}

#define BRO_OP_DO_I(OP, s) \
	SET_U32<s>(rs); \
	BRO_##OP(res, rs); \
	printf("  %s %d: ", #OP, (u32)rs); PRINT_BRANCH(res, true);

#define BRO_OP_DO_M(OP, s) \
	SET_M(rs, s); \
	BRO_##OP(res, rs); \
	printf("  %s %s: ", #OP, #s); PRINT_BRANCH(res, true);

#define TEST_BRO_FUNC(OP) \
BRO_OP_FUNC(OP); \
static void test_##OP() { \
	u32 res; \
	register u128 rs; \
	 \
	printf("%s:\n", #OP); \
	 \
	BRO_OP_DO_I(OP, 0); \
	BRO_OP_DO_I(OP, 1); \
	BRO_OP_DO_I(OP, 2); \
	BRO_OP_DO_I(OP, 0xFFFFFFFF); \
	BRO_OP_DO_I(OP, 0x7FFFFFFF); \
	BRO_OP_DO_I(OP, 0x80000000); \
	 \
	BRO_OP_DO_M(OP, C_ZERO); \
	BRO_OP_DO_M(OP, C_ONE); \
	BRO_OP_DO_M(OP, C_NEGONE); \
	BRO_OP_DO_M(OP, C_S16_MAX); \
	BRO_OP_DO_M(OP, C_S16_MIN); \
	BRO_OP_DO_M(OP, C_S32_MAX); \
	BRO_OP_DO_M(OP, C_S32_MIN); \
	BRO_OP_DO_M(OP, C_S64_MAX); \
	BRO_OP_DO_M(OP, C_S64_MIN); \
	BRO_OP_DO_M(OP, C_GARBAGE1); \
	BRO_OP_DO_M(OP, C_GARBAGE2); \
	\
	printf("\n"); \
}

#define TEST_BO_FUNC(OP) \
BO_OP_FUNC(OP); \
static void test_##OP() { \
	u32 res; \
	 \
	printf("%s:\n", #OP); \
	 \
	BO_##OP(res); \
	printf("  %s: ", #OP); PRINT_BRANCH(res, true); \
	\
	printf("\n"); \
}

#define TEST_BRR_FUNC(OP) \
BRR_OP_FUNC(OP); \
static void test_##OP() { \
	u32 res; \
	 \
	printf("%s:\n", #OP); \
	 \
	BRR_##OP(res); \
	printf("  %s: ", #OP); PRINT_BRANCH(res, true); \
	\
	printf("\n"); \
}

#define TEST_BR_FUNC(OP) \
BR_OP_FUNC(OP); \
static void test_##OP() { \
	u32 res; \
	 \
	printf("%s:\n", #OP); \
	 \
	BR_##OP(res); \
	printf("  %s: ", #OP); PRINT_BRANCH(res, true); \
	\
	printf("\n"); \
}

TEST_BRRO_FUNC(beq);
TEST_BRRO_FUNC(beql);
TEST_BRO_FUNC(bgez);
TEST_BRO_FUNC(bgezal);
TEST_BRO_FUNC(bgezall);
TEST_BRO_FUNC(bgezl);
TEST_BRO_FUNC(bgtz);
TEST_BRO_FUNC(bgtzl);
TEST_BRO_FUNC(blez);
TEST_BRO_FUNC(blezl);
TEST_BRO_FUNC(bltz);
TEST_BRO_FUNC(bltzal);
TEST_BRO_FUNC(bltzall);
TEST_BRO_FUNC(bltzl);
TEST_BRRO_FUNC(bne);
TEST_BRRO_FUNC(bnel);
TEST_BO_FUNC(j);
TEST_BO_FUNC(jal);
TEST_BRR_FUNC(jalr);
TEST_BR_FUNC(jr);

int main(int argc, char **argv) {
	printf("-- TEST BEGIN\n");

	test_beq();
	test_beql();
	test_bgez();
	test_bgezal();
	test_bgezall();
	test_bgezl();
	test_bgtz();
	test_bgtzl();
	test_blez();
	test_blezl();
	test_bltz();
	test_bltzal();
	test_bltzall();
	test_bltzl();
	test_bne();
	test_bnel();
	test_j();
	test_jal();
	test_jalr();
	test_jr();

	printf("-- TEST END\n");

	return 0;
}
