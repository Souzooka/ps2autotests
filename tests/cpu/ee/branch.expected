-- TEST BEGIN
beq:
  beq 0, 0: followed, ignored ra, ran delay slot
  beq 0, 1: skipped, ignored ra, ran delay slot
  beq 1, 1: followed, ignored ra, ran delay slot
  beq 1, 0: skipped, ignored ra, ran delay slot
  beq 2, 2: followed, ignored ra, ran delay slot
  beq -1, 1: skipped, ignored ra, ran delay slot
  beq -1, -1: followed, ignored ra, ran delay slot
  beq C_ZERO, C_ZERO: followed, ignored ra, ran delay slot
  beq C_ZERO, C_ONE: skipped, ignored ra, ran delay slot
  beq C_ONE, C_ZERO: skipped, ignored ra, ran delay slot
  beq C_ONE, C_ONE: followed, ignored ra, ran delay slot
  beq C_ONE, C_NEGONE: skipped, ignored ra, ran delay slot
  beq C_S16_MAX, C_S16_MAX: followed, ignored ra, ran delay slot
  beq C_S16_MIN, C_S16_MIN: followed, ignored ra, ran delay slot
  beq C_S32_MAX, C_S32_MAX: followed, ignored ra, ran delay slot
  beq C_S32_MIN, C_S32_MIN: followed, ignored ra, ran delay slot
  beq C_S64_MAX, C_S64_MAX: followed, ignored ra, ran delay slot
  beq C_S64_MIN, C_S64_MIN: followed, ignored ra, ran delay slot
  beq C_GARBAGE1, C_GARBAGE2: skipped, ignored ra, ran delay slot

beql:
  beql 0, 0: followed, ignored ra, ran delay slot
  beql 0, 1: skipped, ignored ra, skipped delay slot
  beql 1, 1: followed, ignored ra, ran delay slot
  beql 1, 0: skipped, ignored ra, skipped delay slot
  beql 2, 2: followed, ignored ra, ran delay slot
  beql -1, 1: skipped, ignored ra, skipped delay slot
  beql -1, -1: followed, ignored ra, ran delay slot
  beql C_ZERO, C_ZERO: followed, ignored ra, ran delay slot
  beql C_ZERO, C_ONE: skipped, ignored ra, skipped delay slot
  beql C_ONE, C_ZERO: skipped, ignored ra, skipped delay slot
  beql C_ONE, C_ONE: followed, ignored ra, ran delay slot
  beql C_ONE, C_NEGONE: skipped, ignored ra, skipped delay slot
  beql C_S16_MAX, C_S16_MAX: followed, ignored ra, ran delay slot
  beql C_S16_MIN, C_S16_MIN: followed, ignored ra, ran delay slot
  beql C_S32_MAX, C_S32_MAX: followed, ignored ra, ran delay slot
  beql C_S32_MIN, C_S32_MIN: followed, ignored ra, ran delay slot
  beql C_S64_MAX, C_S64_MAX: followed, ignored ra, ran delay slot
  beql C_S64_MIN, C_S64_MIN: followed, ignored ra, ran delay slot
  beql C_GARBAGE1, C_GARBAGE2: skipped, ignored ra, skipped delay slot

bgez:
  bgez 0: followed, ignored ra, ran delay slot
  bgez 1: followed, ignored ra, ran delay slot
  bgez 2: followed, ignored ra, ran delay slot
  bgez -1: skipped, ignored ra, ran delay slot
  bgez 2147483647: followed, ignored ra, ran delay slot
  bgez -2147483648: skipped, ignored ra, ran delay slot
  bgez C_ZERO: followed, ignored ra, ran delay slot
  bgez C_ONE: followed, ignored ra, ran delay slot
  bgez C_NEGONE: skipped, ignored ra, ran delay slot
  bgez C_S16_MAX: followed, ignored ra, ran delay slot
  bgez C_S16_MIN: skipped, ignored ra, ran delay slot
  bgez C_S32_MAX: followed, ignored ra, ran delay slot
  bgez C_S32_MIN: skipped, ignored ra, ran delay slot
  bgez C_S64_MAX: followed, ignored ra, ran delay slot
  bgez C_S64_MIN: skipped, ignored ra, ran delay slot
  bgez C_GARBAGE1: followed, ignored ra, ran delay slot
  bgez C_GARBAGE2: skipped, ignored ra, ran delay slot

bgezal:
  bgezal 0: followed, set ra, ran delay slot
  bgezal 1: followed, set ra, ran delay slot
  bgezal 2: followed, set ra, ran delay slot
  bgezal -1: skipped, set ra, ran delay slot
  bgezal 2147483647: followed, set ra, ran delay slot
  bgezal -2147483648: skipped, set ra, ran delay slot
  bgezal C_ZERO: followed, set ra, ran delay slot
  bgezal C_ONE: followed, set ra, ran delay slot
  bgezal C_NEGONE: skipped, set ra, ran delay slot
  bgezal C_S16_MAX: followed, set ra, ran delay slot
  bgezal C_S16_MIN: skipped, set ra, ran delay slot
  bgezal C_S32_MAX: followed, set ra, ran delay slot
  bgezal C_S32_MIN: skipped, set ra, ran delay slot
  bgezal C_S64_MAX: followed, set ra, ran delay slot
  bgezal C_S64_MIN: skipped, set ra, ran delay slot
  bgezal C_GARBAGE1: followed, set ra, ran delay slot
  bgezal C_GARBAGE2: skipped, set ra, ran delay slot

bgezall:
  bgezall 0: followed, set ra, ran delay slot
  bgezall 1: followed, set ra, ran delay slot
  bgezall 2: followed, set ra, ran delay slot
  bgezall -1: skipped, set ra, skipped delay slot
  bgezall 2147483647: followed, set ra, ran delay slot
  bgezall -2147483648: skipped, set ra, skipped delay slot
  bgezall C_ZERO: followed, set ra, ran delay slot
  bgezall C_ONE: followed, set ra, ran delay slot
  bgezall C_NEGONE: skipped, set ra, skipped delay slot
  bgezall C_S16_MAX: followed, set ra, ran delay slot
  bgezall C_S16_MIN: skipped, set ra, skipped delay slot
  bgezall C_S32_MAX: followed, set ra, ran delay slot
  bgezall C_S32_MIN: skipped, set ra, skipped delay slot
  bgezall C_S64_MAX: followed, set ra, ran delay slot
  bgezall C_S64_MIN: skipped, set ra, skipped delay slot
  bgezall C_GARBAGE1: followed, set ra, ran delay slot
  bgezall C_GARBAGE2: skipped, set ra, skipped delay slot

bgezl:
  bgezl 0: followed, ignored ra, ran delay slot
  bgezl 1: followed, ignored ra, ran delay slot
  bgezl 2: followed, ignored ra, ran delay slot
  bgezl -1: skipped, ignored ra, skipped delay slot
  bgezl 2147483647: followed, ignored ra, ran delay slot
  bgezl -2147483648: skipped, ignored ra, skipped delay slot
  bgezl C_ZERO: followed, ignored ra, ran delay slot
  bgezl C_ONE: followed, ignored ra, ran delay slot
  bgezl C_NEGONE: skipped, ignored ra, skipped delay slot
  bgezl C_S16_MAX: followed, ignored ra, ran delay slot
  bgezl C_S16_MIN: skipped, ignored ra, skipped delay slot
  bgezl C_S32_MAX: followed, ignored ra, ran delay slot
  bgezl C_S32_MIN: skipped, ignored ra, skipped delay slot
  bgezl C_S64_MAX: followed, ignored ra, ran delay slot
  bgezl C_S64_MIN: skipped, ignored ra, skipped delay slot
  bgezl C_GARBAGE1: followed, ignored ra, ran delay slot
  bgezl C_GARBAGE2: skipped, ignored ra, skipped delay slot

bgtz:
  bgtz 0: skipped, ignored ra, ran delay slot
  bgtz 1: followed, ignored ra, ran delay slot
  bgtz 2: followed, ignored ra, ran delay slot
  bgtz -1: skipped, ignored ra, ran delay slot
  bgtz 2147483647: followed, ignored ra, ran delay slot
  bgtz -2147483648: skipped, ignored ra, ran delay slot
  bgtz C_ZERO: skipped, ignored ra, ran delay slot
  bgtz C_ONE: followed, ignored ra, ran delay slot
  bgtz C_NEGONE: skipped, ignored ra, ran delay slot
  bgtz C_S16_MAX: followed, ignored ra, ran delay slot
  bgtz C_S16_MIN: skipped, ignored ra, ran delay slot
  bgtz C_S32_MAX: followed, ignored ra, ran delay slot
  bgtz C_S32_MIN: skipped, ignored ra, ran delay slot
  bgtz C_S64_MAX: followed, ignored ra, ran delay slot
  bgtz C_S64_MIN: skipped, ignored ra, ran delay slot
  bgtz C_GARBAGE1: followed, ignored ra, ran delay slot
  bgtz C_GARBAGE2: skipped, ignored ra, ran delay slot

bgtzl:
  bgtzl 0: skipped, ignored ra, skipped delay slot
  bgtzl 1: followed, ignored ra, ran delay slot
  bgtzl 2: followed, ignored ra, ran delay slot
  bgtzl -1: skipped, ignored ra, skipped delay slot
  bgtzl 2147483647: followed, ignored ra, ran delay slot
  bgtzl -2147483648: skipped, ignored ra, skipped delay slot
  bgtzl C_ZERO: skipped, ignored ra, skipped delay slot
  bgtzl C_ONE: followed, ignored ra, ran delay slot
  bgtzl C_NEGONE: skipped, ignored ra, skipped delay slot
  bgtzl C_S16_MAX: followed, ignored ra, ran delay slot
  bgtzl C_S16_MIN: skipped, ignored ra, skipped delay slot
  bgtzl C_S32_MAX: followed, ignored ra, ran delay slot
  bgtzl C_S32_MIN: skipped, ignored ra, skipped delay slot
  bgtzl C_S64_MAX: followed, ignored ra, ran delay slot
  bgtzl C_S64_MIN: skipped, ignored ra, skipped delay slot
  bgtzl C_GARBAGE1: followed, ignored ra, ran delay slot
  bgtzl C_GARBAGE2: skipped, ignored ra, skipped delay slot

blez:
  blez 0: followed, ignored ra, ran delay slot
  blez 1: skipped, ignored ra, ran delay slot
  blez 2: skipped, ignored ra, ran delay slot
  blez -1: followed, ignored ra, ran delay slot
  blez 2147483647: skipped, ignored ra, ran delay slot
  blez -2147483648: followed, ignored ra, ran delay slot
  blez C_ZERO: followed, ignored ra, ran delay slot
  blez C_ONE: skipped, ignored ra, ran delay slot
  blez C_NEGONE: followed, ignored ra, ran delay slot
  blez C_S16_MAX: skipped, ignored ra, ran delay slot
  blez C_S16_MIN: followed, ignored ra, ran delay slot
  blez C_S32_MAX: skipped, ignored ra, ran delay slot
  blez C_S32_MIN: followed, ignored ra, ran delay slot
  blez C_S64_MAX: skipped, ignored ra, ran delay slot
  blez C_S64_MIN: followed, ignored ra, ran delay slot
  blez C_GARBAGE1: skipped, ignored ra, ran delay slot
  blez C_GARBAGE2: followed, ignored ra, ran delay slot

blezl:
  blezl 0: followed, ignored ra, ran delay slot
  blezl 1: skipped, ignored ra, skipped delay slot
  blezl 2: skipped, ignored ra, skipped delay slot
  blezl -1: followed, ignored ra, ran delay slot
  blezl 2147483647: skipped, ignored ra, skipped delay slot
  blezl -2147483648: followed, ignored ra, ran delay slot
  blezl C_ZERO: followed, ignored ra, ran delay slot
  blezl C_ONE: skipped, ignored ra, skipped delay slot
  blezl C_NEGONE: followed, ignored ra, ran delay slot
  blezl C_S16_MAX: skipped, ignored ra, skipped delay slot
  blezl C_S16_MIN: followed, ignored ra, ran delay slot
  blezl C_S32_MAX: skipped, ignored ra, skipped delay slot
  blezl C_S32_MIN: followed, ignored ra, ran delay slot
  blezl C_S64_MAX: skipped, ignored ra, skipped delay slot
  blezl C_S64_MIN: followed, ignored ra, ran delay slot
  blezl C_GARBAGE1: skipped, ignored ra, skipped delay slot
  blezl C_GARBAGE2: followed, ignored ra, ran delay slot

bltz:
  bltz 0: skipped, ignored ra, ran delay slot
  bltz 1: skipped, ignored ra, ran delay slot
  bltz 2: skipped, ignored ra, ran delay slot
  bltz -1: followed, ignored ra, ran delay slot
  bltz 2147483647: skipped, ignored ra, ran delay slot
  bltz -2147483648: followed, ignored ra, ran delay slot
  bltz C_ZERO: skipped, ignored ra, ran delay slot
  bltz C_ONE: skipped, ignored ra, ran delay slot
  bltz C_NEGONE: followed, ignored ra, ran delay slot
  bltz C_S16_MAX: skipped, ignored ra, ran delay slot
  bltz C_S16_MIN: followed, ignored ra, ran delay slot
  bltz C_S32_MAX: skipped, ignored ra, ran delay slot
  bltz C_S32_MIN: followed, ignored ra, ran delay slot
  bltz C_S64_MAX: skipped, ignored ra, ran delay slot
  bltz C_S64_MIN: followed, ignored ra, ran delay slot
  bltz C_GARBAGE1: skipped, ignored ra, ran delay slot
  bltz C_GARBAGE2: followed, ignored ra, ran delay slot

bltzal:
  bltzal 0: skipped, set ra, ran delay slot
  bltzal 1: skipped, set ra, ran delay slot
  bltzal 2: skipped, set ra, ran delay slot
  bltzal -1: followed, set ra, ran delay slot
  bltzal 2147483647: skipped, set ra, ran delay slot
  bltzal -2147483648: followed, set ra, ran delay slot
  bltzal C_ZERO: skipped, set ra, ran delay slot
  bltzal C_ONE: skipped, set ra, ran delay slot
  bltzal C_NEGONE: followed, set ra, ran delay slot
  bltzal C_S16_MAX: skipped, set ra, ran delay slot
  bltzal C_S16_MIN: followed, set ra, ran delay slot
  bltzal C_S32_MAX: skipped, set ra, ran delay slot
  bltzal C_S32_MIN: followed, set ra, ran delay slot
  bltzal C_S64_MAX: skipped, set ra, ran delay slot
  bltzal C_S64_MIN: followed, set ra, ran delay slot
  bltzal C_GARBAGE1: skipped, set ra, ran delay slot
  bltzal C_GARBAGE2: followed, set ra, ran delay slot

bltzall:
  bltzall 0: skipped, set ra, skipped delay slot
  bltzall 1: skipped, set ra, skipped delay slot
  bltzall 2: skipped, set ra, skipped delay slot
  bltzall -1: followed, set ra, ran delay slot
  bltzall 2147483647: skipped, set ra, skipped delay slot
  bltzall -2147483648: followed, set ra, ran delay slot
  bltzall C_ZERO: skipped, set ra, skipped delay slot
  bltzall C_ONE: skipped, set ra, skipped delay slot
  bltzall C_NEGONE: followed, set ra, ran delay slot
  bltzall C_S16_MAX: skipped, set ra, skipped delay slot
  bltzall C_S16_MIN: followed, set ra, ran delay slot
  bltzall C_S32_MAX: skipped, set ra, skipped delay slot
  bltzall C_S32_MIN: followed, set ra, ran delay slot
  bltzall C_S64_MAX: skipped, set ra, skipped delay slot
  bltzall C_S64_MIN: followed, set ra, ran delay slot
  bltzall C_GARBAGE1: skipped, set ra, skipped delay slot
  bltzall C_GARBAGE2: followed, set ra, ran delay slot

bltzl:
  bltzl 0: skipped, ignored ra, skipped delay slot
  bltzl 1: skipped, ignored ra, skipped delay slot
  bltzl 2: skipped, ignored ra, skipped delay slot
  bltzl -1: followed, ignored ra, ran delay slot
  bltzl 2147483647: skipped, ignored ra, skipped delay slot
  bltzl -2147483648: followed, ignored ra, ran delay slot
  bltzl C_ZERO: skipped, ignored ra, skipped delay slot
  bltzl C_ONE: skipped, ignored ra, skipped delay slot
  bltzl C_NEGONE: followed, ignored ra, ran delay slot
  bltzl C_S16_MAX: skipped, ignored ra, skipped delay slot
  bltzl C_S16_MIN: followed, ignored ra, ran delay slot
  bltzl C_S32_MAX: skipped, ignored ra, skipped delay slot
  bltzl C_S32_MIN: followed, ignored ra, ran delay slot
  bltzl C_S64_MAX: skipped, ignored ra, skipped delay slot
  bltzl C_S64_MIN: followed, ignored ra, ran delay slot
  bltzl C_GARBAGE1: skipped, ignored ra, skipped delay slot
  bltzl C_GARBAGE2: followed, ignored ra, ran delay slot

bne:
  bne 0, 0: skipped, ignored ra, ran delay slot
  bne 0, 1: followed, ignored ra, ran delay slot
  bne 1, 1: skipped, ignored ra, ran delay slot
  bne 1, 0: followed, ignored ra, ran delay slot
  bne 2, 2: skipped, ignored ra, ran delay slot
  bne -1, 1: followed, ignored ra, ran delay slot
  bne -1, -1: skipped, ignored ra, ran delay slot
  bne C_ZERO, C_ZERO: skipped, ignored ra, ran delay slot
  bne C_ZERO, C_ONE: followed, ignored ra, ran delay slot
  bne C_ONE, C_ZERO: followed, ignored ra, ran delay slot
  bne C_ONE, C_ONE: skipped, ignored ra, ran delay slot
  bne C_ONE, C_NEGONE: followed, ignored ra, ran delay slot
  bne C_S16_MAX, C_S16_MAX: skipped, ignored ra, ran delay slot
  bne C_S16_MIN, C_S16_MIN: skipped, ignored ra, ran delay slot
  bne C_S32_MAX, C_S32_MAX: skipped, ignored ra, ran delay slot
  bne C_S32_MIN, C_S32_MIN: skipped, ignored ra, ran delay slot
  bne C_S64_MAX, C_S64_MAX: skipped, ignored ra, ran delay slot
  bne C_S64_MIN, C_S64_MIN: skipped, ignored ra, ran delay slot
  bne C_GARBAGE1, C_GARBAGE2: followed, ignored ra, ran delay slot

bnel:
  bnel 0, 0: skipped, ignored ra, skipped delay slot
  bnel 0, 1: followed, ignored ra, ran delay slot
  bnel 1, 1: skipped, ignored ra, skipped delay slot
  bnel 1, 0: followed, ignored ra, ran delay slot
  bnel 2, 2: skipped, ignored ra, skipped delay slot
  bnel -1, 1: followed, ignored ra, ran delay slot
  bnel -1, -1: skipped, ignored ra, skipped delay slot
  bnel C_ZERO, C_ZERO: skipped, ignored ra, skipped delay slot
  bnel C_ZERO, C_ONE: followed, ignored ra, ran delay slot
  bnel C_ONE, C_ZERO: followed, ignored ra, ran delay slot
  bnel C_ONE, C_ONE: skipped, ignored ra, skipped delay slot
  bnel C_ONE, C_NEGONE: followed, ignored ra, ran delay slot
  bnel C_S16_MAX, C_S16_MAX: skipped, ignored ra, skipped delay slot
  bnel C_S16_MIN, C_S16_MIN: skipped, ignored ra, skipped delay slot
  bnel C_S32_MAX, C_S32_MAX: skipped, ignored ra, skipped delay slot
  bnel C_S32_MIN, C_S32_MIN: skipped, ignored ra, skipped delay slot
  bnel C_S64_MAX, C_S64_MAX: skipped, ignored ra, skipped delay slot
  bnel C_S64_MIN, C_S64_MIN: skipped, ignored ra, skipped delay slot
  bnel C_GARBAGE1, C_GARBAGE2: followed, ignored ra, ran delay slot

j:
  j: followed, ignored ra, ran delay slot

jal:
  jal: followed, set ra, ran delay slot

jalr:
  jalr: followed, set ra, ran delay slot

jr:
  jr: followed, ignored ra, ran delay slot

-- TEST END
